Addweightbalance1

sts1skj opened the pull request:
Added weight-balancing to weight brushes
    
This is the first of a sequence of changes to add automatic weight balancing to Outfit Studio.  This change just affects the three weight brushes.  With this change, the weights for the bones on a vertex will be maintained to have more or less the same sum as they did before.
    
A lot of the changes in this commit are designed to support the next changes, which will add balance-bone selection to the GUI.

Please let me know if you'd prefer that I do one huge pull request for all of the weight-balancing stuff rather than doing it incrementally like this.  I thought you would prefer finer control and feedback as I add features one by one.

sts1skj committed:
Added weight-balancing to weight brushes

This is the first of a sequence of commits to add automatic weight
balancing to Outfit Studio.  This commit just affects the three weight
brushes.  With this change, the weights for the bones on a vertex will
be maintained to have more or less the same sum as they did before.

A lot of the changes in this commit are designed to support the next
commit, which will add balance-bone selection to the GUI.

sts1skj committed:
Undid tiny bit of previous weight-balancing commit

I accidentally changed something that shouldn't have been changed.
I changed it back.

ousnius commented:
@sts1skj Separate pull requests where possible are preferred so I can more easily see and understand the changes.

This seems fine but I'd like a way to turn this live normalization off. There are some editing cases where you don't want it to normalize before you're done doing your thing. I assume the way it was implemented now the old logic of just adding the paint isn't there anymore.

When painting everything to one bone (100% red), this happens to the other weights that previously affected the area (they don't fully become zero).
![image](https://user-images.githubusercontent.com/7034795/73399134-03fcd000-42e7-11ea-930c-62ec70da11ab.png)

sts1skj committed:
Tweaked weight-brush calculations so weights will zero more consistently

The previous iteration of the weight brush didn't fully zero other weights
when maximizing the selected bone's weights.  I've tweaked the numerics
so it's better at zeroing things.

sts1skj committed:
Made sure weight painting works when total weight is zero.

Fixed the code so the selected bone's weight can be increased to one when all
other bone weights are zero.

sts1skj commented:
I've tweaked the numerics so hopefully the other weights will go to zero more consistently.  I've also fixed a bug that was preventing weight from increasing when all other weights were zero.

As for wanting the option to not normalize, here was my plan for the user interface:

I would add an icon next to each bone (similar to the visibility icon for shapes) that would indicate "balance bone" (on) or "not balance bone" (off).  Bones would default to off.  If no bones are balance bones, all non-selected bones are used as balance bones (are included in the normalization).  If only a few non-selected bones are balance bones, weight would only be subtracted from those bones' weights.

In this scheme, if you want to change the weight of a bone without it being normalized, pick one or more balance bones that are already zero for that area.  Then no bone weights would change other than the selected bone's.

When I originally suggested this feature, I used the term "complementary bone".  I like the term "balance bone" better now.  If you can think of a better term, please tell me it.

ousnius commented:
I would probably just actually use "Normalized Bones", because that's what it does. Complementary and balance aren't really explaining what it actually does, so one can just use the real technical term (normalization) that at least people familiar with 3D modelling or some math might be familiar with.

Zeroing out works well now. 👍

sts1skj commented:
@ousnius Are you waiting for me to do something?

ousnius commented:
@sts1skj Sorry, I was busy. I will merge but before a future release of a new version I will have to go through and test everything thoroughly again.
